### ISSUE TITLE
URIUtil.canonicalPath incorrectly handles paths with query parameters containing '../'

### ISSUE THREAD

**Issue by ontology-rory** [eclipse/jetty.project#1079]
When preforming some penetration testing of our web app, we discovered several NullPointerExceptions in the log with pages getting 500 errors. We are using Jetty 9.3.8

`
java.lang.NullPointerException
	at org.eclipse.jetty.server.Response.sendRedirect(Response.java:671)
	at org.eclipse.jetty.server.Response.sendRedirect(Response.java:691)
	at org.apache.wicket.protocol.http.servlet.ServletWebResponse.sendRedirect(ServletWebResponse.java:297)
`

This was due to URIUtil.canonicalPath returning NULL for the path "/foo/bar?10&5=../../../../../../../../../../../../../../../../etc/passwd".
I would have expected the path to be returned AS IS, as the ../ sequence are in the query parameters rather than the query info. For example

- /foo/bar?10&5=../../../../../../../../../../../../../../../../etc/passwd => /foo/bar?10&5=../../../../../../../../../../../../../../../../etc/passwd
- /foo/../bar?10&5=../../../../../../../../../../../../../../../../etc/passwd => /bar?10&5=../../../../../../../../../../../../../../../../etc/passwd

**Comment by joakime** [comment]
@gregw should this be fixed by making `URIUtil.canonicalPath()` query string aware?
Or by making change locally in `Response.sendRedirect()` only, applying the `URIUtil.canonicalPath()` to the path portion only (ignoring the query portion)

**Comment by joakime** [comment]
This is trivially easy to replicate.

Just add the following test condition to the `URIUtilTest#testCanonicalPath()` use cases

```java
{"/foo/bar?10&5=../../../../etc/passwd", "/foo/bar?10&5=../../../../etc/passwd"},
```

**Comment by gregw** [comment]
I'll look at the rfc early next week. I'm not sure if / is legal in a query
string and it may need to be encoded.

But returning null is wrong. It should be left as is or a malformed uri
exception from canonical path.

Send redirect should also better handle a null return.

**Comment by joakime** [comment]
The query portion can contain "/" unencoded.

See https://tools.ietf.org/html/rfc3986#section-3.4

**Comment by gregw** [comment]
Fixed by #1487 and #1480
